File: 20_07_July/0731_17680_캐시/Solution.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {

    public int solution(int cacheSize, String[] cities) {
        Queue<String> queue = new LinkedList<>();
        int cost = 0;

        for (String city : cities) {
            String lower = city.toLowerCase();

            if (queue.contains(lower)) {
                cost += 1;
                queue.remove(lower);
            } else {
                cost += 5;
            }

            queue.add(lower);

            if (queue.size() > cacheSize) {
                queue.poll();
            }
        }

        return cost;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        System.out.println(solution.solution(3, new String[] {"Jeju", "Pangyo", "Seoul", "NewYork", "LA", "Jeju", "Pangyo", "Seoul", "NewYork", "LA"}));
        System.out.println(solution.solution(0, new String[] {"A", "A", "A", "A"}));
    }
}

File: 20_07_July/0731_17680_캐시/Solution.py
from collections import deque

def solution(cacheSize, cities):
    deq = deque()
    cost = 0
    s = set()

    # 0이면 loop 계산할 필요가 없음.
    if cacheSize == 0:
        return len(cities) * 5

    for city in cities:
        name = city.lower()

        if name in s:
            cost += 1
            deq.remove(name)
        else:
            cost += 5
            if len(deq) == cacheSize:
                s.remove(deq.popleft())
            
            s.add(name)

        deq.append(name)
        

    return cost